### PR TITLE
Support numpy vertical

### PR DESCRIPTION
**Pull request by edumugi**
Implement changes discussed in #9360.

@rongou @trivialfis

### CODE REVIEW

File: src/c_api/c_api.cc [hunk 1]
```diff
@@ -464,7 +464,10 @@ XGB_DLL int XGDMatrixCreateFromDense(char const *data,
   float missing = GetMissing(config);
   auto n_threads = OptionalArg<Integer, int64_t>(config, "nthread", 0);
   xgboost_CHECK_C_ARG_PTR(out);
-  *out = new std::shared_ptr<DMatrix>(DMatrix::Create(&adapter, missing, n_threads));
+  auto data_split_mode =
```
Comment (rongou) [inline src/c_api/c_api.cc:467]: This line should be above the check on `out`.

Comment (edumugi) [reply src/c_api/c_api.cc:466]: Fixed!

File: src/c_api/c_api.cc [hunk 2]
```diff
@@ -463,8 +463,11 @@ XGB_DLL int XGDMatrixCreateFromDense(char const *data,
   auto config = Json::Load(StringView{c_json_config});
   float missing = GetMissing(config);
   auto n_threads = OptionalArg<Integer, int64_t>(config, "nthread", 0);
+  auto data_split_mode =
+      static_cast<DataSplitMode>(OptionalArg<Integer, int64_t>(config, "data_split_mode", 0));
   xgboost_CHECK_C_ARG_PTR(out);
-  *out = new std::shared_ptr<DMatrix>(DMatrix::Create(&adapter, missing, n_threads));
+  *out = new std::shared_ptr<DMatrix>(
+      DMatrix::Create(&adapter, missing, n_threads, std::string(""), data_split_mode));
```
Comment (rongou) [inline src/c_api/c_api.cc:470]: You can just pass in the empty string `""` here, no need to construct a `std::string`.

**Review by rongou — APPROVED** [review]

File: src/c_api/c_api.cc (hunk 2 above)
Comment (edumugi) [reply src/c_api/c_api.cc]: Is it OK like this?

**Review by trivialfis — COMMENTED** [review]
@rongou Any recommendations for writing tests?

**Comment by rongou** [comment]
Not sure. We haven't been testing column-split data loading on the python side. Perhaps we can add a method to `DMatrix`, something like `get_data_split_mode`, and verify it after we construct the numpy DMatrix.

**Comment by trivialfis** [comment]
Got it, I will leave it for now.

**Review by trivialfis — APPROVED** [review]

**Pull request merged**